src/arc.rs

```diff
@@ -184,6 +176,30 @@ impl<T: ?Sized> Arc<T> {
         this.as_ptr()
     }
 
+    /// Reconstruct the `Arc<T>` from a raw pointer obtained from into_raw()
+    ///
+    /// Note: This raw pointer will be offset in the allocation and must be preceded
+    /// by the atomic count.
+    ///
+    /// It is recommended to use OffsetArc for this
+    ///
+    ///  # Safety
+    /// - The given pointer must be a valid pointer to `T` that came from [`Arc::into_raw`].
+    /// - After `from_raw`, the pointer must not be accessed.
+    #[inline]
+    pub unsafe fn from_raw(ptr: *const T) -> Self {
+        // To find the corresponding pointer to the `ArcInner` we need
+        // to subtract the offset of the `data` field from the pointer.
+
+        // SAFETY: `ptr` comes from `ArcInner.data`, so it must be initialized.
+        let offset_of_data = ArcInner::<T>::offset_of_data(ptr);
+
+        // SAFETY: `ptr` points to `ArcInner.data`,
+        //   so subtraction results in the beginning of the `ArcInner`.
```

Comment:
I think I still want an explicit line here saying `from_raw_inner` accepts a pointer to the beginning of the allocation, not the data part.

This is something that is more confusing about this code and while we haven't been great about

Reply:
https://github.com/unicode-org/icu4x/pull/4785

Reply:
Another version!